### ISSUE TITLE
java.lang.LinkageError thrown at startup after updating from 1.7 to 1.9

### ISSUE THREAD

**Issue by gsmirnov-splk** [open-telemetry/opentelemetry-java-instrumentation#5155]
**Describe the bug**
After updating splunk-otel-java from 1.5.0 to 1.6.1 (comes with `opentelemetry-java-instrumentation` updated to 1.9), the instrumented application (Apache Druid) fails to start with this error:

```
java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for name: "org/jboss/netty/channel/ChannelFutureListener"
```

After initial troubleshooting, @iNikem suggested that I open an issue in this repo. Subsequent troubleshooting confirmed that the vendor-neutral upstream `opentelemetry-javaagent - version: 1.10.0` also fails with the same error.

**Steps to reproduce**
1. Make sure you have Java 8 selected
2. Run Apache Druid quickstart: (Steps 1&2 [here](https://druid.apache.org/docs/latest/tutorials/index.html))
3. Ensure that the Historical is up and running, `http://localhost:8083/status` should return a JSON. It may take a minute to start
4. Download a [splunk-otel-java agent](https://github.com/signalfx/splunk-otel-java/releases/download/v1.6.1/splunk-otel-javaagent-all.jar) and place it into the `apache-druid-0.22.1` folder
4. Attach the agent to the historical node: `echo "-javaagent:splunk-otel-javaagent-all.jar" >> conf/druid/single-server/micro-quickstart/historical/jvm.config`
5. Start druid again: `./bin/start-micro-quickstart`

**What did you expect to see?**
Historical starts up like it did before.

**What did you see instead?**
Historical fails to start:
```
[Mon Jan 17 13:47:01 2022] Running command[historical], logging to[apache-druid-0.22.1/var/sv/historical.log]: bin/run-druid historical conf/druid/single-server/micro-quickstart
[Mon Jan 17 13:47:15 2022] Command[historical] exited (pid = 2747, exited = 1)
[Mon Jan 17 13:47:15 2022] Command[historical] failed, see logfile for more details: apache-druid-0.22.1/var/sv/historical.log
```
```
2022-01-17T11:46:41,236 ERROR [main] org.apache.druid.cli.CliHistorical - Error when starting up.  Failing.
java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for name: "org/jboss/netty/channel/ChannelFutureListener"
```

**What version are you using?**
`opentelemetry-javaagent - version: splunk-1.6.1-otel-1.9.2`
`opentelemetry-javaagent - version: 1.10.0`

**Environment**
JVM: Amazon Coretto 1.8.0_312; BellSoft 1.8.0_302
OS: docker on multiple different host operating systems; WSL2

**Additional context**
Attached are two log files with Agent (`-Dotel.javaagent.debug=true`) and JVM (`-XX:+TraceClassLoading`) debug information: `success.log` (agent version 1.5.0 attached) and `fail.log` (agent version 1.6.1 attached). The former loads the `org.jboss.netty.channel.ChannelFutureListener` class just once after transformation:

```
[otel.javaagent 2022-01-14 19:21:45:936 +0000] [main] DEBUG io.opentelemetry.javaagent.tooling.AgentInstaller$TransformLoggingListener - Transformed org.jboss.netty.channel.ChannelFutureListener -- sun.misc.Launcher$AppClassLoader@18b4aac2
[Loaded org.jboss.netty.channel.ChannelFutureListener from file:/opt/druid/lib/netty-3.10.6.Final.jar]
```

The latter indeed loads it twice, one time before transforming:

```
[Loaded org.jboss.netty.channel.ChannelFutureListener from file:/opt/druid/lib/netty-3.10.6.Final.jar]
[otel.javaagent 2022-01-14 19:15:29:088 +0000] [main] DEBUG
 io.opentelemetry.javaagent.tooling.AgentInstaller$TransformLoggingListener - Transformed org.jboss.netty.channel.ChannelFutureListener -- sun.misc.Launcher$AppClassLoader@18b4aac2
[Loaded org.jboss.netty.channel.ChannelFutureListener from file:/opt/druid/lib/netty-3.10.6.Final.jar]
java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for name: "org/jboss/netty/channel/ChannelFutureListener"
```

---
**Attachments:**
* [success.log](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/7881636/success.log) with `splunk-1.5.0-otel-1.7.0`
* [fail.log](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/7881637/fail.log) with `splunk-1.6.1-otel-1.9.2`

**Comment by laurit** [comment]
Loading `ChannelFutureListener` triggers injecting helper classes. As there is a helper (`ConnectionListener`) that depends on `ChannelFutureListener` defining it also loads `ChannelFutureListener`. Now when the original load of `ChannelFutureListener` tries to define the class it is already loaded and fails with `java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for name: "org/jboss/netty/channel/ChannelFutureListener"`
It is possible to circumvent this by running with `-noverify`

**Comment by gsmirnov-splk** [comment]
@laurit how did it work before then, were the affected classes not instrumented at all? Or was the instrumentation logic different for inter-class dependencies? I suppose the instrumentation should be able to get the info it needs from the class file without actually loading it into the application's classloader.

I tried running with `-noverify` locally and indeed, it no longer crashes at startup. But this is NOT an option I will ever enable in production and certainly the customers cannot be asked to disable the verified either.

**Comment by laurit** [comment]
@gsmirnov-splk 
> how did it work before then, were the affected classes not instrumented at all? 

The instrumentation was slightly different. To trigger this bug instrumentation needs to be triggered by a class that is super type of a helper class. `ConnectionListener` that implements `ChannelFutureListener` was not present in the version that didn't have this issue.

> I tried running with `-noverify` locally and indeed, it no longer crashes at startup. But this is NOT an option I will ever enable in production and certainly the customers cannot be asked to disable the verified either.

This wasn't really meant as a suggestion. By just looking at druid source it is not obvious why `ChannelFutureListener` is loaded from. It was meant as hint that it is probably loaded by the verifier to others who decide to have a go at this.

**Comment by gsmirnov-splk** [comment]
@breedx-splk if you help me get a `.jar` that I can pass as a `-javaagent:`, I'll be happy to verify.

**Comment by breedx-splk** [comment]
> @breedx-splk if you help me get a `.jar` that I can pass as a `-javaagent:`, I'll be happy to verify.

Thanks! The release is here https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/download/v1.10.0/opentelemetry-javaagent.jar

**Comment by mateuszrzeszutek** [comment]
Here's an idea, not sure if it'll work: instead of instrumenting `ChannelFutureListener`, in `NettyChannelInstrumentation` we could make the instrumented method return a custom `ChannelFuture` that would decorate the added listeners with `parentContext.makeCurrent()` and simply delegate all other methods. This would remove the matcher on `ChannelFutureListener` and hopefully fix that issue.

**Comment by gsmirnov-splk** [comment]
@breedx-splk yeah, I can confirm too -- the error reproduces with `opentelemetry-javaagent - version: 1.10.0`

**Comment by trask** [comment]
@gsmirnov-splk thanks for reporting this! it's now fixed in the [1.10.1 patch release](https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v1.10.1)

**Comment by gsmirnov-splk** [comment]
I can confirm that with 1.10.1, the issue no longer reproduces. Apache Druid starts successfully with the agent attached again.

Thanks, everyone!
